### ISSUE TITLE
torch.fx.Interpreter modules don't get compiled

### ISSUE THREAD

**Issue by suo** [pytorch/pytorch#107631]
As a small example:
```
import torch
import torch.fx

class InterpreterModule(torch.fx.GraphModule):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.interpreter = torch.fx.Interpreter(self)

    def __call__(self, *args, **kwargs):
        return self.interpreter.run(*args, enable_io_processing=False)

class Mod(torch.nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, x):
        x = x + x
        x = x * x
        return x

gm = torch.fx.symbolic_trace(Mod())
imodule = InterpreterModule({}, gm.graph)
opt_mod = torch.compile(imodule)

for i in range(1000):
    opt_mod(torch.ones(2, 3))
```
Turning on the logging, I see that we avoid compiling over `torch.fx.Interpreter` because it's getting skipfile'd. When I try adding `torch.fx.Interpreter` to the skipfile, it skips again with `GraphModule.graph`, so I gave up here and decided to record an issue instead.

Not sure how easy/hard this is to support with dynamo, but it would be nice. Generally we like using `torch.fx.Interpreter` instead of the torch.fx Python codegen for two reasons:
1. You get better stack traces in the case of exceptions
2. It's easier to debug/interact with intermediate

**Comment by ezyang** [comment]
One downside to compiling through Interpreter directly is that you cannot graph break, it will kick you out to the top level of the loop. If you generated the Python codegen this would work.

**Comment by suo** [comment]
Yeah, we prefer `torch.fx.Interpreter` for the two reasons I mentioned at the bottom, but at least one of them (stack traces) could be made to work with Python codegen. It's a plausible alternative to making compiling Interpreter.

Although—the graph break thing doesn't seem like such a big deal to me. We mostly use it for models that already have no graph breaks (e.g. exported models).

**Comment by anijain2305** [comment]
Assigning to myself for figuring how much work this is.

**Comment by anijain2305** [comment]
My initial impression of this is - it requires some work.

The main objective here is that Dynamo should not cause a graph break in the Interpreter `run` (and the children frames). Otherwise, each node will end up in a separate Dynamo fx graph. Some of the hard graph breaks are

1) custom objects with `__iter__` - this happens in `_node_list` class
2) no graph break skip on context manager - this happens in `run_node`
3) Possible exceptions - which I don't think are well supported in Dynamo

So, this would require some planning. Will chat offline to come up with a plan.
